jdk/test/sun/nio/cs/TestEBCDICLineFeed.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2017, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8186803
+ * @summary Check if the linefeed is de/encoded correctly in ebcdic
+ * @modules jdk.charsets
+ */
+
+public class TestEBCDICLineFeed {
+
+    public static void main(String[] args) throws Exception {
+
+        String[] csnames = new String[] {
+            "IBM037",  "IBM1025", "IBM1026", "IBM1112", "IBM1122", "IBM1123",
+            "IBM1166", "IBM273",  "IBM277",  "IBM278",  "IBM280",  "IBM284",
+            "IBM285",  "IBM297",  "IBM420",  "IBM424",  "IBM500",  "IBM838",
+            "IBM870",  "IBM871",   "IBM875", "IBM918",  "IBM930",  "IBM935",
+            "IBM937",  "IBM939",
+
+            "IBM01140", "IBM01141", "IBM01142", "IBM01143", "IBM01144",
+            "IBM01145", "IBM01146", "IBM01147", "IBM01148", "IBM01149",
+        };
+
+        int errs = 0;
+        for (String cs : csnames) {
+            // No JDK-8186803 in 8u so allow u+000A -> 0x25
+            boolean jdk8186803 = cs.startsWith("IBM0114");
+            byte[] bb = "\n".getBytes(cs);
+            if (bb.length != 1 ||
+               (bb[0] != 0x15 && !jdk8186803) ||
+               (bb[0] != 0x25 && jdk8186803)) {
+                System.out.printf(" error: %s c2b  u+000A -> %x%n",
+                                  cs,  bb[0] & 0xff);
+                errs++;
+            }
```

Comment:
Took some thinking, but I think this is okay.

Are we sure the old (pre-8186803) translations are always "U+000A" -> 25? I am not sure, it looked to me like it could be either mapped to E15 or E25.

Reply:
The changes I've made to detect IBM0114x were necessary to get the test to pass. Prior to that, it was failing because those character sets were producing 0x25.

Reply:
As to including this test, it seemed appropriate. The intention is for this backport to add tests to ensure the status quo, rather than change any behaviour (as Severin notes). So this version of `TestEBCDICLineFeed.java` is ensuring that the 8186803 bug is present in 8u.

I agree about it being a bit of a brainteaser. It was a pain to extract the 8186803 changes from the rest and I was surprised that even removing the `.nr` file additions still led to a mismatch in `TestCharsetMappings`, fixed by the hack. As you've seen, the IBM114x maps map both 0x15 and 0x25 to U+000a, so going the other way can produce either. In `TestCharsetMappings`, both are produced but 0x15 succeeds without the hack (i.e. `eq` is already true). As far as I can work out. the `.nr` files (no round trip) stop the 0x25 conversion happening, so only 0x15 occurs.